packages/contractkit/src/identity/offchain/schemas/utils.ts

```diff
@@ -0,0 +1,348 @@
+    [wrapper.self, ...toAddresses].map(async (toAddress) =>
+      distributeSymmetricKey(wrapper, dataPath, fetchKey.result, toAddress)
+    )
+  )
+}
+
+/**
+ * Reads and decrypts a symmetric key that has been encrypted to your
+ * data encryption key.
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param senderAddress the address that encrypted this key to you
+ */
+const readSymmetricKey = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  senderAddress: string
+): Promise<Result<Buffer, SchemaErrors>> => {
+  const accounts = await wrapper.kit.contracts.getAccounts()
+  const wallet = wrapper.kit.getWallet()
+  const [readerPubKey, senderPubKey] = await Promise.all([
+    accounts.getDataEncryptionKey(wrapper.self),
+    accounts.getDataEncryptionKey(senderAddress),
+  ])
+
+  if (readerPubKey === null) {
+    return Err(new UnavailableKey(senderAddress))
+  }
+
+  const readerPublicKeyAddress = publicKeyToAddress(readerPubKey)
+  if (!wallet.hasAccount(readerPublicKeyAddress)) {
+    return Err(new UnavailableKey(readerPublicKeyAddress))
+  }
+
+  const sharedSecret = await wallet.computeSharedSecret(readerPublicKeyAddress, senderPubKey)
+  const computedDataPath = getCiphertextLabel(
+    `${dataPath}.key`,
+    sharedSecret,
+    senderPubKey,
+    readerPubKey
+  )
+  const encryptedPayload = await wrapper.readDataFromAsResult(
+    senderAddress,
+    (buf) => buildEIP712TypedData(wrapper, computedDataPath, buf),
+    computedDataPath
+  )
+
+  if (!encryptedPayload.ok) {
+    return Err(new OffchainError(encryptedPayload.error))
+  }
+
+  const payload = await wallet.decrypt(readerPublicKeyAddress, encryptedPayload.result)
+  return Ok(payload)
+}
+
+/**
+ * Reads and decrypts a payload that has been encrypted to your data encryption key. Will
+ * resolving the symmetric key used to encrypt the payload.
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param senderAddress the address that encrypted this key to you
+ */
+export const readEncrypted = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  senderAddress: string
+): Promise<Result<Buffer, SchemaErrors>> => {
+  const encryptedPayloadPath = `${dataPath}.enc`
+  const [payload, key] = await Promise.all([
+    wrapper.readDataFromAsResult(
+      senderAddress,
+      (buf) => buildEIP712TypedData(wrapper, encryptedPayloadPath, buf),
+      encryptedPayloadPath
+    ),
+    readSymmetricKey(wrapper, dataPath, senderAddress),
+  ])
+
+  if (!payload.ok) {
+    return Err(new OffchainError(payload.error))
+  }
+  if (!key.ok) {
+    return Err(key.error)
+  }
+
+  if (key.result.length !== KEY_LENGTH) {
+    return Err(new InvalidKey())
+  }
+
+  return Ok(
+    AES128Decrypt(key.result, payload.result.slice(0, IV_LENGTH), payload.result.slice(IV_LENGTH))
```

Comment:
This is inconsistent with what we do for encryption above, which is written using only the Node `crypto` APIs. It does seems to be the same, but is a bit confusing. Id recommend either splitting the `AES128EncryptAndHMAC` function in `ecies.ts`to create an `AES128Encypt` function, or implement this section using the Node API (possibly just copy-pasting the code from `ecies.ts`) to make this match.

Reply:
I think making an `AES128Encrypt` function makes sense, one question, should the new `AES128Encrypt` and `AES128Decrypt` functions take the `iv` parameter? 

The other option is that `AES128Encrypt` always generates a new one, `AES128Decrypt` always looks for the `iv` as the first 16 bytes.

Reply:
I think that makes sense, if not I will revert

Reply:
Option 2 is the way to go. the IV is generally considered part of the ciphertext in most crypto abstractions, and it's a detail that doesn't need to be controlled by the user of a high-level library.